File: main/java/control/CommandDelegation.java
package control;

import units.Rover;

import java.util.Arrays;
import java.util.Optional;

public enum CommandDelegation {

    RIGHT("R") {
        public Rover delegate(Rover rover) {return rover.right();}
    },
    LEFT("L") {
        public Rover delegate(Rover rover) { return rover.left(); }
    },
    MOVE("M") {
        public Rover delegate(Rover rover) { return rover.move();}
    };


    private String command;

    CommandDelegation(String command) {
        this.command = command;
    }

    public static CommandDelegation fromString(String string) {

        Optional<CommandDelegation> delegator = Arrays
                .stream(values())
                .filter(d -> d.command.equals(string))
                .findFirst();

        if (!delegator.isPresent()) {
            throw new IllegalArgumentException("No such command");
        }

        return delegator.get();
    }

    public abstract Rover delegate(Rover rover);

}

File: main/java/com/marsrovers/control/CommandDelegation.kt
package com.marsrovers.control

import com.marsrovers.units.Rover

import java.util.Arrays
import java.util.Optional

enum class CommandDelegation private constructor(private val command: String) {

    RIGHT("R") {
        override fun delegate(rover: Rover): Rover = rover.right()
    },
    LEFT("L") {
        override fun delegate(rover: Rover): Rover = rover.left()
    },
    MOVE("M") {
        override fun delegate(rover: Rover): Rover = rover.move()
    };

    abstract fun delegate(rover: Rover): Rover

    companion object {

        fun fromString(string: String): CommandDelegation {

            val delegation = Arrays
                    .stream(values())
                    .filter { d -> d.command == string }
                    .findFirst()

            if (!delegation.isPresent) {
                throw IllegalArgumentException("No such command " + string)
            }

            return delegation.get()
        }
    }

}
